### ISSUE TITLE
problem with overloaded function and generics

### ISSUE THREAD

**Issue by cleoold** [microsoft/pyright#2194]
**Describe the bug**
when you define an overloaded function, and pass this function to another function that accepts generic parameters, the following error is produced:

**To Reproduce**
Type code

**Expected behavior**
mypy does not produce this error.

**Screenshots or Code**
```python
@overload
def ff1(real: float): ...  # type: ignore
@overload
def ff1(real: str): ...

def fun(f: Callable[[T], Any], p: T):
    return f(p)

fun(ff1, 4)
fun(ff1, '4')
~~~~~~~~~ ^
          error: Argument of type "Literal['4']" cannot be assigned to parameter "p" of type "T@fun" in function "fun"
            Type "float" cannot be assigned to type "str" (reportGeneralTypeIssues)
```

there is no error if you change the order of `f` and `p`

**VS Code extension or command-line**
pyright 1.1.162

**Additional context**
this valid code is flagged as error and may be relevant:
```py
map(complex, input().split(' '))
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
No overloads for "__init__" match the provided arguments
  Argument types: (Type[complex], list[str]) (reportGeneralTypeIssues)

# typeshed:
class complex:
    @overload
    def __new__(cls: Type[_T], real: float = ..., imag: float = ...) -> _T: ...
    @overload
    def __new__(cls: Type[_T], real: str | SupportsComplex | SupportsIndex | complex) -> _T: ...
    ....
class map(Iterator[_S], Generic[_S]):
    @overload
    def __init__(self, __func: Callable[[_T1], _S], __iter1: Iterable[_T1]) -> None: ...
    ....
```

**Comment by erictraut** [comment]
Thanks for the bug report. These two issues were related. Both of them will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.174, which I just published. It will also be included in the next release of pylance.
